### ISSUE TITLE
User configurable shortcuts

### ISSUE THREAD

**Issue by letsfindaway** [letsfindaway/OpenBoard#11]
The ability to configure the shortcuts to trigger actions is one of the often requested features for OpenBoard. In this issue I will collect some of the ideas for implementing this.

An example of a shortcut configuration dialog can be seen in QtCreator. Here a screenshot of the relevant part:
![image](https://user-images.githubusercontent.com/35591413/109409597-64e85300-7994-11eb-8f58-6331575e63e5.png)
The dialog consists of a table of all actions, grouped in topics. For each action, the name, a description and the currently assigned shortcut is listed. When a row is selected, the shortcut is shown in a text input field below the table. Here it can be literally edited, or the key combination can be recorded. There is also a button to reset the shortcut to the application default. Below the table, the dialog also offers the option to import or export shortcut definitions.

Not shown on the screenshot: When a shortcut was modified, it is shown in bold italic. During entry, it is checked for correctness and a warning message is shown below the input field. Here also a message is shown when an entry collides with other shortcut definitions.

Special ideas for OpenBoard:

- Limit the shortcuts to a single key. No key sequences with more than one key.
- Checkbox to enable activation of actions without `Ctrl` key, i.e. `E` switches to eraser, not `Ctrl+E`.
- Additional column to assign special mouse or stylus buttons to the actions.

Questions coming into my mind:

- How do we group the actions?
- Should the actions be scoped? I.e. for document mode, board mode etc? Note that the shortcut context won't help here, as all actions are assigned to the main window. We would have to move the actions to distinct widgets.
- Shall the shortcuts be saved as part of the normal user configuration file? Or should there be a special shortcut configuration file, which also resembles the import/export format?
- Could we generate a printable list of effective shortcuts?

**Comment by letsfindaway** [comment]
Some more implementation considerations:

- The UI handling is implemented in the `UBPreferencesController`.
- The `UBPreferencesController` allows to record a key combination, a mouse button (not the left and right) and a stylus button to an action.
- There is a new `UBShortcutManager` which keeps a list of actions and manages the associated shortcuts. This class also assigns mouse and stylus buttons to the actions and checks for duplicate assignments.
- Information beyond the active keyboard shortcuts, e.g. the original default shortcut and the mouse and stylus button assignments are kept with the actions using dynamic properties.
- The `UBShortcutManager` exposes its information as a `QAbstractTableModel`, which is used by the `UBPreferencesController`.
- Keyboard shortcuts are handled via the standard means of the `QAction`.
- Mouse and stylus button events are intercepted by the `UBApplication`. If the `UBPreferencesController` is in recording mode, then these events are recorded. Else if the `UBShortcutManager` has an associated action, that action is triggered. In other cases, the default handling of the event is invoked.

**Comment by letsfindaway** [comment]
As this is not a new topic, I want to collect what I can find in existing (open and closed) issues and pull requests:

- Keyboard shortcuts, https://github.com/OpenBoard-org/OpenBoard/issues/82 is the most active issue here. It requests documentation and configuration of shortcuts in the preferences section.
- Hotkey for color switch, https://github.com/OpenBoard-org/OpenBoard/issues/213. This issues proposes a feature to cycle through the colors via a stylus button.
- Map tablet buttons to common actions, https://github.com/OpenBoard-org/OpenBoard/issues/378
- Using the mouse forward and backward keys as undo and redo, https://github.com/OpenBoard-org/OpenBoard/issues/440 wants to map mouse buttons to actions.
- Allow configurable Shortcuts, https://github.com/OpenBoard-org/OpenBoard/pull/321. This pull request adds additional shortcut functions and a simple configuration page in the preferences. It did not get it's way to the master or development branch.

There are much more, mainly closed as duplicates. Let me categorize them as far as I understand them:

1. Most requested is the ability to configure keyboard shortcuts for all actions.
2. Also very common is the request to use special mouse buttons or stylus buttons to trigger actions.
3. Here often also new actions are proposed like cycling through colors.

In this issue I want to concentrate on the first two: make the shortcuts configurable and also use mouse and stylus buttons to trigger actions. Number three requires new actions, which will not be the primary purpose of this issue.

**Comment by letsfindaway** [comment]
Here is the initial draft of a user interface for configurable shortcuts:

![image](https://user-images.githubusercontent.com/35591413/110519335-79071f80-810d-11eb-9876-34908bc7dcd7.png)

- The table lists actions sorted by groups.
- There is a line edit field to filter the list. The entered text is searched in the first two columns. Only matching actions along with their group titles will be shown.
- Each action may have a key sequence, a mouse button and a stylus button specified. They are listed in colums 3 to 5.
- A checkbox below the table allows to activate a shortcut without pressing the `Ctrl` key. So e.g. `I` is sufficient to select the pen instead of `Ctrl+I`.
- The bottom part is an edit area for the selected table entry. Shortcuts are "recorded", i.e. you just activate the proper keyboard sequence, mouse button or stylus button, and these actions are recorded and will then be applied as shortcuts to the action. This prevents entering typos like `Crtl+X` or button specifiers which simply do not exist on the user's hardware. Note that the "recording area" will most probably not be present in the final implementation, as it might not be necessary.
- An error message will be shown in a status line at the bottom, when a shortcut is duplicated. The table will then automatically be filtered to show the offending entry.
- There is the option to revert a single action or the complete table to defaults.

@kaamui: I would really appreciate your feedback - and that of any other user, too!

**Comment by kaamui** [comment]
I'll try to look at it asap. It looks great ! Thank you a lot for the help you provide to this project.

**Comment by letsfindaway** [comment]
> I'll try to look at it asap. It looks great ! Thank you a lot for the help you provide to this project.

Thanks for your feedback. Currently the code is not complete, but I want to check that I'm going in the right direction. I will inform you when I have pushed something working.

I'm currently also thinking whether I could also include a feature to **_configure_** keys and buttons which modify tool behavior when pressed, as this was also often requested (e.g. panning, switching to eraser, etc). I will for sure not implement those features, but at least a framework to configure the keys and buttons used for such things if they will come in the future.

**Comment by letsfindaway** [comment]
Further TODOs:

- [ ] Add all relevant actions to the `UBShortcutManager` and group them.
- [ ] Implement filtering.
- [ ] Collision detection and warning.
- [ ] Persistence using the `UBSettings`.
- [ ] Shortcuts without `Ctrl` including collision detection.

**Comment by letsfindaway** [comment]
@kaamui The current status at https://github.com/letsfindaway/OpenBoard/tree/shortcut-configuration are ready to give you an impression of the current status of my work. What is missing are the two unchecked boxes in the comment above. Especially the second will require some effort.

**Comment by kaamui** [comment]
>  Fix and improve conflict detection for "Ignore Ctrl" mode.
About this particular point, where the focus needs to be for the shortcut to be triggered ? Is it automatically handled by Qt that when focus is on an editable QGraphicsTextItem it should not be triggered ? Are the events "attached" to a global handler that is listening or does it need to be attached to every view ?

**Comment by letsfindaway** [comment]
tl;dr: it works!

Longer explanation: 

I'm not the Qt event handling expert, but what I understood from the documentation and all my tests is the following:

- The key event is created by Qt and delivered to the `QApplication`.
- Here you could filter the event, but that's not what I'm doing for keyboard events.
- The event then reaches the `event()` function of the `QApplication`. Here it is decided, to whom the event is offered.
- First the utmost active leaf in the widget tree is selected.
- If the event is not handled here, it bubbles upward towards the root.
- At each level, shortcut handlers attached to that level are also considered.
- If nobody handles the event, it is dropped.

Currently most actions are attached to buttons. Therefore shortcuts only work, when the buttons are visible. I'm additionally attaching the actions to the main window. So the shortcuts are also working when the associated button is not visible, because visibility of the main window is sufficient.

The option "Ignore Ctrl" adds secondary shortcuts to the actions. So an action with shortcut "Ctrl+B" gets a secondary shortcut "B". What happens when the `QGraphicsTextItem` is active? Just check the event handling list above: The key press is first offered to the `QGraphicsTextItem` before it is evaluated as a possible shortcut. As the widget handles the event, it will not trigger the shortcut action.

I think the important fact is, that the shortcut check for actions attached to the main window is only executed when the key event was not handled by an active child widget before.

Please correct me if I'm wrong, but this is what I have understood.

**Comment by kaamui** [comment]
Hi,

Thanks for your work, and the time you take to answer. I'll try to look at your work as soon as I can.

**Comment by letsfindaway** [comment]
Instead of the `UBModifierButton` mentioned above I will implement a "switch tool while key is pressed" feature. So if you're e.g. using the pen tool and are pressing and holding down Ctrl+E (default shortcut for eraser tool), then the tool will switch to "Eraser" as long as you keep the keys pressed. When the E is released, the previous tool (here: pen tool) is selected again.

This will work with all shortcuts of the tool palette and in addition to the function where pressing a shortcut key combination switches permanently to a new tool. The distinguishing of the behavior is based on time: When the keys are pressed longer than the keyboard autorepeat interval, then the temporary switching will be activated.

**Comment by letsfindaway** [comment]
Closed as PR was merged
